### PR TITLE
Fix detection of scp-style URIs to support non-standard SSH ports

### PR DESCRIPTION
**Pull request by ksonj**
This PR attempts to fix the broken detection of SCP style URIs (e.g. `git@github.com:NixOS/nix`). Currently the regex also matches URIs with a fixed port, e.g. `ssh://git@github.com:22/NixOS/nix`. This worked with Nix 2.3 but does not currently with Nix 2.4

Since scp URIs do not have a scheme this simply adds a test for the existence of a "://" substring

### CODE REVIEW

**Comment by ksonj** [comment]
I just realized there's already https://github.com/NixOS/nix/issues/5376 and a very similar fix in https://github.com/NixOS/nix/pull/5465. The only difference is that the other PR assumes a scheme of "ssh://" to rule out scp-style which I believe is not always accurate.

E.g. the URL `https://ksonj@github.com:443/NixOS/nix` works with Nix 2.3 and with this fix, but not with 2.4 and still not with the other PR

File: src/libexpr/primops/fetchTree.cc [hunk 1]
```diff
@@ -74,8 +74,11 @@ std::string fixURI(std::string uri, EvalState & state, const std::string & defau
 
 std::string fixURIForGit(std::string uri, EvalState & state)
 {
+    /* Detects scp-style uris (e.g. git@github.com:NixOS/nix) and fixes
+     * them by removing the `:` and assuming a scheme of `ssh://`
+     * */
     static std::regex scp_uri("([^/].*)@(.*):(.*)");
```
Comment (regnat) [inline src/libexpr/primops/fetchTree.cc:80]: Couldn’t this be fixed by just changing the regex to

```suggestion
    static std::regex scp_uri("([^/]*)@(.*):(.*)");
```

?

I have the feeling that this `.` was just a typo

Comment (regnat) [reply src/libexpr/primops/fetchTree.cc:80]: (I think that the best specification for what’s allowed in an ssh username is https://datatracker.ietf.org/doc/html/rfc2396#section-3.2.2 , which fobids `/` in a username. So it’s not overly restrictive either)

Comment (ksonj) [reply src/libexpr/primops/fetchTree.cc]: :bulb: that is true. Let me quickly fix that

**Review by regnat — APPROVED** [review]

**Comment by Ericson2314** [comment]
Should this be backported if there is a regression from 2.3? Do we have a test?

**Comment by regnat** [comment]
> Should this be backported if there is a regression from 2.3?

Yup’ indeed, it should

> Do we have a test?

Unfortunately not. I couldn’t find a way to write a meaningful test for this without either some non-trivial refactoring to make this pure and allow for unit-testing, or having to spawn-up an ssh server to make sure that everything works end-to-end, which might be a bit too much

**Comment by regnat** [comment]
> or having to spawn-up an ssh server to make sure that everything works end-to-end, which might be a bit too much

Although now that I think of it, that could be a good fit for a hydra test

**Pull request merged**